### PR TITLE
proposed fix 821 Range-select in Song Editor

### PR DESCRIPTION
**Pull request by curlymorphic**
Proposed for issue #821 Range-select in Song Editor
The TimeLine widget checks to see if's parent is a SongEditor. if so Shift Click can be used on the time line to select all TCO in the time range.

I used the current TrackContainerView's rubber band to ensure consistency for selection code. 
Added 2 functions to TrackContainerView, 
  selectRegionFromPixels, the sets the rubberband using the x,y pixels
  StopRubberBand, that is used to remover the rubberband an mouseRelease

### CODE REVIEW

**Comment by Sti2nd** [comment]
Is this drag on timeline with Shift or between loop markers, or something else?

**Comment by curlymorphic** [comment]
Shift drag, this was suggest by @diizy in the original issue #821

**Comment by tresf** [comment]
@Sti2nd try shift in a file explorer window.  That is likely the behavior he is describing.

**Comment by tresf** [comment]
Thanks for clarification. :)

**Comment by Sti2nd** [comment]
Can you implement this in Piano Roll as well?

**Comment by curlymorphic** [comment]
> Can you implement this in Piano Roll as well?

yes ofc, it was listed as a separate issue #446 , so I am treating it as a such,

**Comment by Sti2nd** [comment]
Nice :)

**Comment by Lukas-W** [comment]
> The TimeLine widget checks to see if's parent is a SongEditor.

I think that's not a good way to design this feature. The `TimeLine` widget should rather provide a `setRegionSelectAccepted(bool)` method and emit a signal like `regionSelected(int x_from, int x_to)` (or using time instead of pixels). The SongEditor could then use this signal.

**Comment by curlymorphic** [comment]
> I think that's not a good way to design this feature. The TimeLine widget should rather provide a setRegionSelectAccepted(bool) method and emit a signal like regionSelected(int x_from, int x_to) (or using time instead of pixels). The SongEditor could then use this signal.

I like this idea :)

**Comment by diizy** [comment]
On 01/20/2015 08:26 PM, Lukas W wrote:
>
>     The TimeLine widget checks to see if's parent is a SongEditor.
>
> I think that's not a good way to design this feature. The |TimeLine|
> widget should rather provide a |setRegionSelectAccepted(bool)| method
> and emit a signal like |regionSelected(int x_from, int x_to)| (or
> using time instead of pixels). The SongEditor could then use this signal.
>

This is a good example of an acceptable use of signals: we're not
dealing with any DSP code here, the signal is emitted from a one-time
action (not a periodical one) and received by another UI element (song
editor). I support the idea.

**Comment by curlymorphic** [comment]
Changed this to use sockets and slots.

File: src/gui/TimeLineWidget.cpp [hunk 1]
```diff
@@ -39,6 +39,7 @@
 #include "NStateButton.h"
 #include "GuiApplication.h"
 #include "TextFloat.h"
+#include "SongEditor.h"
```
Comment (Lukas-W) [inline src/gui/TimeLineWidget.cpp]: That line can be removed now, right? :)

**Comment by Lukas-W** [comment]
Thanks @curlymorphic! :+1:

**Pull request merged**